docs/source/coding_guidelines.md

```diff
@@ -0,0 +1,253 @@
+# Design and Coding Guidelines
+
+## Design Phase
+
+For non-trivial changes, please communicate your plans early on to reach maximum impact and to avoid duplicate work.
+Create an [issue](https://github.com/microsoft/hi-ml/issues) or a
+[discussion](https://github.com/microsoft/hi-ml/discussions) that describes the problem that you are trying to solve,
+and describe why this adds a positive contribution to the project.
+
+Tag one of the core `hi-ml` team (@ant0nsc, @mebristo, @fepegar) to get input on your problem.
+In particular, you should find out if this has already been worked on or thought about.
+This will help you to get to know about existing code fragments that can simplify your task, and integrate better with existing code.
+
+We recommend to work with a core `hi-ml` team member to create a design for your work.
+Depending on the size of the task at hand, a design can be anywhere between a short description of your plan in a GitHub discussion or issue, or it can be a shared document where you list your design options and invite feedback.
+Your design partner will later also be well positioned to review your pull request (PR) because they will be already familiar with your plan.
+
+## Setting up your development environment
+
+Please see our [detailed instructions](developers.md).
+
+## Coding Guidelines
+
+### Naming
+
+Please follow the general [PEP 8](https://peps.python.org/pep-0008/) Python rules for naming:
+
+* Variables, function and method names should use `snake_case` (lower case with under scores separating words)
+* Class names should follow `PascalCase` (first letter of each word capitalized).
+
+To improve readability, functions or methods that return boolean values should follow a `is_...`, `has_...`, `use...` pattern, like `is_status_ok` instead of `ok`.
+
+### Static Analysis and Linting
+
+We use `flake8` as a linter, and `mypy` and pyright for static typechecking. Both tools run as part of the PR workflow, and must run without errors for a contribution to be accepted.
+`mypy` requires that all functions and methods carry type annotations.
+See the [`mypy` documentation](https://mypy.readthedocs.io/en/latest/getting_started.html#function-signatures-and-dynamic-vs-static-typing) for more information.
+
+We highly recommend to run all those tools _before_ pushing the latest changes to a PR.
+If you have `make` installed, you can run both tools in one go via `make check` (from the repository root folder).
+
+### String Style
```

Comment:
Should this section be within the 'documentation' header? But also, is this not a bit excessive for a research codebase?

Reply:
@fepegar has been adding comments along those lines to previous PRs, that's why I put it in (if it's complained about, there should be a written reason for the complaint). If it was just for me: I would not have put it in. Happy to do a majority vote on that one.

Reply:
I've moved that to documentation examples now, and ditched the one on error messages.